File: api/app/services/sse_broker.py
"""In-memory SSE broker: broadcasts new articles to all connected clients."""

import asyncio
import logging

logger = logging.getLogger(__name__)

CLIENT_QUEUE_MAXSIZE = 64


class SSEBroker:
    """asyncio.Queue-based Pub/Sub broker."""

    def __init__(self) -> None:
        self._queues: set[asyncio.Queue[dict]] = set()

    def subscribe(self) -> asyncio.Queue[dict]:
        """Create and return a new client queue."""
        queue: asyncio.Queue[dict] = asyncio.Queue(maxsize=CLIENT_QUEUE_MAXSIZE)
        self._queues.add(queue)
        return queue

    def unsubscribe(self, queue: asyncio.Queue[dict]) -> None:
        """Remove a client queue on disconnect."""
        self._queues.discard(queue)

    async def broadcast(self, event: dict) -> None:
        """Send event to all connected clients. Drop if queue is full."""
        for queue in list(self._queues):
            try:
                queue.put_nowait(event)
            except asyncio.QueueFull:
                logger.warning("SSE client queue full, dropping event")

    @property
    def client_count(self) -> int:
        return len(self._queues)


article_broker = SSEBroker()

File: api/src/services/sse-broker.ts
import { EventEmitter } from "node:events";
import { rootLogger } from "../lib/logger.js";

const logger = rootLogger.child({ service: "sse-broker" });

const CLIENT_QUEUE_MAXSIZE = 64;
const MAX_SSE_CONNECTIONS = 20;

export class ConnectionLimitExceeded extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ConnectionLimitExceeded";
  }
}

// Any JSON-serializable object; the route stringifies it as the SSE payload.
export type SSEEvent = object;

interface ClientQueue {
  id: number;
  events: SSEEvent[];
  resolve: ((value: SSEEvent) => void) | null;
}

export class SSEBroker {
  private clients = new Map<number, ClientQueue>();
  private nextId = 0;
  private emitter = new EventEmitter();

  subscribe(): number {
    if (this.clients.size >= MAX_SSE_CONNECTIONS) {
      throw new ConnectionLimitExceeded(
        `Maximum SSE connections (${MAX_SSE_CONNECTIONS}) exceeded`,
      );
    }
    this.nextId = (this.nextId + 1) % Number.MAX_SAFE_INTEGER;
    const id = this.nextId;
    this.clients.set(id, { id, events: [], resolve: null });
    return id;
  }

  unsubscribe(clientId: number): void {
    this.clients.delete(clientId);
  }

  broadcast(event: SSEEvent): void {
    for (const [, client] of this.clients) {
      if (client.resolve) {
        // Client is waiting for an event
        const resolve = client.resolve;
        client.resolve = null;
        resolve(event);
      } else if (client.events.length < CLIENT_QUEUE_MAXSIZE) {
        client.events.push(event);
      } else {
        logger.warn("SSE client queue full, dropping event");
      }
    }
  }

  async waitForEvent(clientId: number, timeoutMs = 1000): Promise<SSEEvent | null> {
    const client = this.clients.get(clientId);
    if (!client) return null;

    // Check buffered events first
    if (client.events.length > 0) {
      return client.events.shift() ?? null;
    }

    // Wait for next event
    return new Promise<SSEEvent | null>((resolve) => {
      const timer = setTimeout(() => {
        client.resolve = null;
        resolve(null);
      }, timeoutMs);

      client.resolve = (event) => {
        clearTimeout(timer);
        resolve(event);
      };
    });
  }

  get clientCount(): number {
    return this.clients.size;
  }
}

export const articleBroker = new SSEBroker();
